crates/accelerate/src/permutation.rs

```diff
@@ -77,6 +77,65 @@ fn get_ordered_swap(pattern: &ArrayView1<i64>) -> Vec<(i64, i64)> {
     swaps
 }
 
+fn pattern_to_cycles(pattern: &ArrayView1<i64>, invert_order: &bool) -> Vec<Vec<usize>> {
+    // vector keeping track of which elements in the permutation pattern have been visited
+    let mut explored: Vec<bool> = vec![false; pattern.len()];
+
+    // vector to store the cycles
+    let mut cycles: Vec<Vec<usize>> = Vec::new();
+
+    // cast the input pattern in terms of integers to usize, such that it can be used as index
+    // also invert the bit ordering if ``invert_order`` is true
+    let permutation: Array1<usize> = if *invert_order {
+        invert(pattern) // implies cast to usize
+    } else {
+        // TODO change to using Array1 insted of Vec
+        pattern.mapv(|x| x as usize)
+    };
+
+    for mut i in permutation.clone() {
+        let mut cycle: Vec<usize> = Vec::new();
+
+        // follow the cycle until we reached an entry we saw before
+        while !explored[i] {
+            cycle.push(i);
+            explored[i] = true;
+            i = permutation[i];
+        }
+        // cycles must have more than 1 element
+        if cycle.len() > 1 {
+            cycles.push(cycle);
+        }
+    }
+
+    cycles
+}
+
+/// Given a disjoint cycle decomposition, decomposes every cycle into a SWAP
+fn decompose_cycles(cycles: &Vec<Vec<usize>>) -> Vec<(usize, usize)> {
+    let mut swaps: Vec<(usize, usize)> = Vec::new();
+
+    for cycle in cycles {
+        let length = cycle.len();
+
+        if length > 2 {
+            // handle first element separately, which accesses the last element
+            swaps.push((cycle[length - 1], cycle[length - 3]));
+            for i in 1..(length - 1) / 2 {
+                swaps.push((cycle[i - 1], cycle[length - 3 - i]));
+            }
+        }
+
+        // no size check needed, cycles always have at least 2 elements
+        swaps.push((cycle[length - 1], cycle[length - 2]));
+        for i in 1..length / 2 {
+            swaps.push((cycle[i - 1], cycle[length - 2 - i]));
+        }
+    }
+
+    swaps
```

Comment:
this also feels like it could be done with iterators, but I'm not sure if would be too convoluted. I can think about it a bit more.

Reply:
I was discussing this one with @jakelishman, and since the resulting swaps are not iterated over but returned to Python as list, there's likely not much advantage in turning this into an iterator. However, if we can find a way to make `cycles` an iterator, that might be worthwhile!